OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs

```diff
@@ -0,0 +1,405 @@
+			{ 724, Pair.New("QUAD", "Mercenaries") },
+			{ 725, Pair.New("COMBATO", "Mercenaries") },
+			{ 726, Pair.New("MISSILETANK", "Mercenaries") },
+			{ 727, Pair.New("SIEGETANK", "Mercenaries") },
+			{ 728, Pair.New("CARRYALLO", "Mercenaries") },
+		};
+
+		readonly Ruleset rules;
+		readonly FileStream stream;
+		readonly string tilesetName;
+		readonly TerrainTile clearTile;
+
+		Map map;
+		Size mapSize;
+		TileSet tileSet;
+		List<TerrainTemplateInfo> tileSetsFromYaml;
+
+		D2kMapImporter(string filename, string tileset, Ruleset rules)
+		{
+			tilesetName = tileset;
+			this.rules = rules;
+
+			try
+			{
+				clearTile = new TerrainTile(0, 0);
+				stream = File.OpenRead(filename);
+
+				if (stream.Length == 0 || stream.Length % 4 != 0)
+					throw new Exception("The map is in an unrecognized format!");
+
+				Initialize(filename);
+				FillMap();
+			}
+			catch (Exception e)
+			{
+				Console.WriteLine(e);
+				map = null;
+			}
+			finally
+			{
+				stream.Close();
+			}
+		}
+
+		public static Map Import(string filename, string mod, string tileset, Ruleset rules)
+		{
+			var map = new D2kMapImporter(filename, tileset, rules).map;
+			if (map == null)
+				return null;
+
+			map.RequiresMod = mod;
+			map.MakeDefaultPlayers();
+
+			return map;
+		}
+
+		void Initialize(string mapFile)
+		{
+			mapSize = new Size(stream.ReadUInt16(), stream.ReadUInt16());
+
+			tileSet = rules.TileSets["ARRAKIS"];
+			map = Map.FromTileset(tileSet);
+			map.Title = Path.GetFileNameWithoutExtension(mapFile);
+			map.Author = "Westwood Studios";
+			map.MapSize.X = mapSize.Width + 2 * MapCordonWidth;
+			map.MapSize.Y = mapSize.Height + 2 * MapCordonWidth;
+			map.Bounds = new Rectangle(MapCordonWidth, MapCordonWidth, mapSize.Width, mapSize.Height);
+
+			map.Smudges = Exts.Lazy(() => new List<SmudgeReference>());
+			map.Actors = Exts.Lazy(() => new Dictionary<string, ActorReference>());
+			map.MapResources = Exts.Lazy(() => new CellLayer<ResourceTile>(TileShape.Rectangle, new Size(map.MapSize.X, map.MapSize.Y)));
+			map.MapTiles = Exts.Lazy(() => new CellLayer<TerrainTile>(TileShape.Rectangle, new Size(map.MapSize.X, map.MapSize.Y)));
+
+			map.Options = new MapOptions();
+
+			// Get all templates from the tileset YAML file that have at least one frame and an Image property corresponding to the requested tileset
+			// Each frame is a tile from the Dune 2000 tileset files, with the Frame ID being the index of the tile in the original file
+			tileSetsFromYaml = tileSet.Templates.Where(t => t.Value.Frames != null
+				&& t.Value.Image.ToLower() == tilesetName.ToLower()).Select(ts => ts.Value).ToList();
+		}
+
+		void FillMap()
+		{
+			while (stream.Position < stream.Length)
+			{
+				var tileInfo = stream.ReadUInt16();
+				var tileSpecialInfo = stream.ReadUInt16();
+				var tile = GetTile(tileInfo);
+
+				var locationOnMap = GetCurrentTilePositionOnMap();
+
+				Console.WriteLine(locationOnMap.X + "   " + locationOnMap.Y);
```

Comment:
I'm sure this is left-over from debugging and will be removed before merging, right?